patches/api/0327-Add-CompostItemEvent.patch

```diff
@@ -0,0 +1,139 @@
+new file mode 100644
+index 0000000000000000000000000000000000000000..5c16b8f2ee8536f25dc5e88f3efa9cf8052d638b
+--- /dev/null
++++ b/src/main/java/com/destroystokyo/paper/event/block/CompostItemEvent.java
+@@ -0,0 +1,64 @@
++package com.destroystokyo.paper.event.block;
++
++import org.bukkit.block.Block;
++import org.bukkit.event.HandlerList;
++import org.bukkit.event.block.BlockEvent;
++import org.bukkit.inventory.ItemStack;
++import org.jetbrains.annotations.NotNull;
++
++/**
++ * Called when an item is about to be composted by a hopper.
++ * To prevent hoppers from moving items into composters, cancel the InventoryMoveItemEvent.
++ */
++public class CompostItemEvent extends BlockEvent {
++
++    private static final HandlerList HANDLER_LIST = new HandlerList();
++
++    private ItemStack item;
++    private boolean willRaiseLevel;
++
++    public CompostItemEvent(@NotNull Block composter, @NotNull ItemStack item, boolean willRaiseLevel) {
++        super(composter);
++        this.item = item;
++        this.willRaiseLevel = willRaiseLevel;
++    }
++
++    /**
++     * Gets the item that was used on the composter.
++     *
++     * @return the item
++     */
++    @NotNull
++    public ItemStack getItem() {
++        return this.item;
++    }
++
++    /**
++     * Gets whether or not the composter will rise a level.
++     *
++     * @return true if successful
++     */
++    public boolean willRaiseLevel() {
++        return this.willRaiseLevel;
++    }
++
++    /**
++     * Sets whether or not the composter will rise a level.
++     *
++     * @param willRaiseLevel true if the composter should rise a level
++     */
++    public void setWillRaiseLevel(boolean willRaiseLevel) {
++        this.willRaiseLevel = willRaiseLevel;
++    }
++
++    @Override
++    public @NotNull HandlerList getHandlers() {
++        return HANDLER_LIST;
++    }
++
++    @NotNull
++    public static HandlerList getHandlerList() {
++        return HANDLER_LIST;
++    }
++
++}
+diff --git a/src/main/java/com/destroystokyo/paper/event/entity/EntityCompostItemEvent.java b/src/main/java/com/destroystokyo/paper/event/entity/EntityCompostItemEvent.java
+new file mode 100644
+index 0000000000000000000000000000000000000000..c6a742c2191c3fce4c56658d61686b67eb9307af
+--- /dev/null
++++ b/src/main/java/com/destroystokyo/paper/event/entity/EntityCompostItemEvent.java
+@@ -0,0 +1,57 @@
++package com.destroystokyo.paper.event.entity;
++
++import com.destroystokyo.paper.event.block.CompostItemEvent;
++import org.bukkit.block.Block;
++import org.bukkit.entity.Entity;
++import org.bukkit.event.Cancellable;
++import org.bukkit.event.HandlerList;
++import org.bukkit.inventory.ItemStack;
++import org.jetbrains.annotations.NotNull;
++
++/**
++ * Called when an item is about to be composted by an entity.
++ */
++public class EntityCompostItemEvent extends CompostItemEvent implements Cancellable {
++
++    private static final HandlerList HANDLER_LIST = new HandlerList();
```

Comment:
I wouldn't have this event have its own Handler list. I think CompostItemEvent should fire for *all* composts, not just ones from hoppers.